### ISSUE TITLE
No way to ignore a file in a folder

### ISSUE THREAD

**Issue by petdance** [petdance/ack3#43]
_From @AndrewRayCode on August 6, 2014 0:10_

From http://stackoverflow.com/questions/24815522/how-do-i-make-ack-ignore-a-file-in-a-directory/24824663#24824663

I have a file in a directory `a/b/c.js`

I can't ignore `c.js` with `--ignore-file=match:c.js` because I want to match other file named `c.js`

I can't ignore dir `a/b` with `--ignore-dir=a/b` because **there are other files I care about in `a/b`**

I've tried:

```
--ignore-file=a/b/c.js       "Invalid filter specification"
```

I've tried:

```
--ignore-file=match:/a\/b\/c.js/
```

Doesn't work, and I'm guessing that it's because ack doesn't read the file path as part of the match, just the name.

I've tried:

```
--ignore-dir=match:/a\/b\/c.js/     " Non-is filters are not yet supported for --ignore-dir (what?)"
```

I can't find anything useful in the [Ack manual](http://beyondgrep.com/documentation/ack-2.12-man.html) soup. How do I ignore a file in a directory?

_Copied from original issue: petdance/ack2#479_

**Comment by petdance** [comment]
_From @hoelzro on August 6, 2014 1:45_

@DelvarWorld You're completely right; I don't know if there _is_ a way to do this.  `--ignore-file` and `--ignore-dir` both only act on the last component in a file path (in this case, `c.js`), and you said you _don't_ want to ignore other files with the same name, so that's out.  Interestingly enough, though, is the fact that `-g` regexes apply to the whole path, but `--ignore-*` only apply to the last part of a path...

**Comment by petdance** [comment]
_From @pmorch on October 15, 2014 13:34_

I also find I'm needing this. We have a huge `somedir/Makefile` that is autogenerated. I want to search most Makefiles, just not that one.

What I'd like was to be able to put this in `.ackrc`:

```
--ignore-file-path=match:somedir/Makefile$
```

### Expected Behavior

Ideally it always matches against the full name of the file all the way from `/`, similar to Perl's `$File::Find::fullname`.  So that if I did this, it also ignored `./Makefile`:

```
$ cd /path/to/somedir
$ ack foobar
```

But I could easily live with `./Makefile` not being ignored, if `ack` only matches against the relative path. I'll run most `acks` somewere above `somedir` anyway.

**Comment by petdance** [comment]
See also #7.

**Comment by petdance** [comment]
https://github.com/petdance/ack3/blob/dev/dev/IGNORE.md

**Comment by pmorch** [comment]
@petdance I find that a Scenario 3 is missing from https://github.com/petdance/ack3/blob/dev/dev/IGNORE.md

## Scenario 3
--ignore-dir can ignore paths as well as directories. E.g. `--ignore-dir=foo/bar` ignores any directory, where `foo/dir` is contained in the path, regardless of where ack was started from.

(This of course assumes there is access to the full path during finding; something like `$File::Find::fullname`)

**Comment by petdance** [comment]
--ignore-dir and --ignore-file are not going to change.  The behavior of --ignore-path is up for discussion.

**Comment by pmorch** [comment]
I totally understand. The thing is that https://github.com/petdance/ack3/blob/dev/dev/IGNORE.md itself contains:

> --ignore-dir can ignore paths as well as directories

There is no mention of --ignore-path in IGNORE.md. And I was trying to stay consistent with that. It matters not to me whether a the existing --ignore-dir gets expanded or --ignore-path gets introduced.

**Comment by petdance** [comment]
IGNORE.md was old notes.  I've updated them to current thinking.  Thanks.

**Comment by MrMichaelJKelly** [comment]
Also voting for this feature - I've forked the repo and am willing to do some exploratory dev work on it if that's helpful.

**Comment by petdance** [comment]
I'd love to see `--ignore-path`.  Before anyone starts coding, though, please define the behavior so that we can hash it out.

**Comment by MrMichaelJKelly** [comment]
The right behavior IMHO is for --ignore-path to treat a path differently than a file.  So while the behavior for --ignore-file is to apply the file path to this directory and all sub-directories, for --ignore-path it should be **path specific** - so 

`--ignore-path=./a.txt`
Should ignore a.txt only in the current directory (the ./) while
`--ignore-path=a.txt`

Probably should ignore it in this directory and all sub-directories.

Similarly, given a directory hierarchy like this:

`ROOT
      .ackrc
      DIR1
             .ackrc
             DIR2
                      a.txt
`

`--ignore-path=dir2/a.txt`
in Dir1/.ackrc above should ignore a.txt only in dir2. 

You could, potentially, have something like

`--ignore-path=*/a.txt`

Which would ignore it in all sub-directories, or even

`--ignore-path=*/gen/a.txt`

Which would ignore it only in sub-directories name “gen”

**Comment by petdance** [comment]
What is the "current directory" if the `--ignore-path` is in /etc/ackrc?  What if it's in a user's personal .ackrc?  What if it's in a project .ackrc?  Or can they not appear in there?

--ignore-file and --ignore-dir don't handle globs.  I don't think it's a good idea for --ignore-path to handle globs, either.

What is the specific use case for --ignore-path that we're trying to get at?  With real filenames and directory names for clarity, how would you use --ignore-path today?

**Comment by petdance** [comment]
> The right behavior IMHO is for --ignore-path to treat a path differently than a file.

Yes, absolutely.  But let's start talking about specific use case and why we want this feature.

**Comment by MrMichaelJKelly** [comment]
OK, I agree that we shouldn't make it different than --ignore-file and --ignore-dir - it should be similar to those.

I think the key problem that --ignore-path is trying to solve is that we want to identify *a single file* to be ignored, not a class of files.  That also argues against the glob, which is good.

So assume I have a layout like this:

<pre>
/
  src
     /my_code
           access.hpp
           access.cpp
     /boost
            access.hpp
</pre>

I want the access.hpp in boost to not be included in searches, but if I happen to have code elsewhere (e.g. in my_code in example above) that has a file called access.hpp, I *don't* want to exclude that.

So the way I would fix this in my view with --ignore-path is that I could _either_ put in the global .ackrc:

`--ignore-path=is:/src/boost/access.hpp`

_OR_ in the .ackrc in /src:

`--ignore-path=is:boost/access.hpp`

_OR_ in an .ackrc in /src/boost:

`--ignore-path=is:access.hpp`

All provide different versions of a _full path_ to the file that I want to ignore, and that is the new functionality - I'm identifying an exact group of files that I don't want included in search.

Note that others have asked for this:

- https://stackoverflow.com/questions/24815522/how-do-i-make-ack-ignore-a-file-in-a-directory
- https://stackoverflow.com/questions/35752670/is-there-a-way-to-ignore-file-path-using-ack

# Issues

- Should --ignore-path also take other modifiers, i.e. ext:, match:.  I would say _no_ since it is trying to uniquely identify a path, not a group of files.  But don't want to break consistency.

- Will this be hard to do given current Ack processing, i.e. is it hard to know the full path to the file when you're processing filters like this?

**Comment by petdance** [comment]
> Note that others have asked for this:

You don't have to sell me on the idea of the feature.  It's a just a matter of figuring out a sensible way to do it.

**Comment by MrMichaelJKelly** [comment]
OK - understand.  See if you agree with the approach above.
